Resfeber/CoreData/CoreDataStack.swift

```diff
@@ -9,64 +9,47 @@
 import Foundation
 import CoreData
 
-open class CoreDataStack {
-  public static let modelName = "Trip"
-
-  public static let model: NSManagedObjectModel = {
-    let modelURL = Bundle.main.url(forResource: modelName, withExtension: "momd")!
-    return NSManagedObjectModel(contentsOf: modelURL)!
-  }()
-
-  public init() {
-  }
-
-  public lazy var mainContext: NSManagedObjectContext = {
-    return storeContainer.viewContext
-  }()
-
-  public lazy var storeContainer: NSPersistentContainer = {
-    let container = NSPersistentContainer(name: CoreDataStack.modelName, managedObjectModel: CoreDataStack.model)
-    container.loadPersistentStores { _, error in
-      if let error = error as NSError? {
-        fatalError("Unresolved error \(error), \(error.userInfo)")
-      }
+class CoreDataStack {
+    public static let modelName = "Trip"
+    static let shared = CoreDataStack()
+
+    public static let model: NSManagedObjectModel = {
+        let modelURL = Bundle.main.url(forResource: modelName, withExtension: "momd")!
+        return NSManagedObjectModel(contentsOf: modelURL)!
+    }()
+    
+    private init() {
     }
-    return container
-  }()
-
-  public func newDerivedContext() -> NSManagedObjectContext {
-    let context = storeContainer.newBackgroundContext()
-    return context
-  }
-
-  public func saveContext() {
-    saveContext(mainContext)
-  }
-
-  public func saveContext(_ context: NSManagedObjectContext) {
-    if context != mainContext {
-      saveDerivedContext(context)
-      return
-    }
-
-    context.perform {
-      do {
-        try context.save()
-      } catch let error as NSError {
-        fatalError("Unresolved error \(error), \(error.userInfo)")
-      }
+    
+    public lazy var mainContext: NSManagedObjectContext = {
+        return storeContainer.viewContext
+    }()
+    
+    public lazy var storeContainer: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: CoreDataStack.modelName, managedObjectModel: CoreDataStack.model)
+        container.loadPersistentStores { _, error in
+            if let error = error as NSError? {
+                fatalError("Unresolved error \(error), \(error.userInfo)")
+            }
+        }
+        return container
+    }()
+    
+    public func newDerivedContext() -> NSManagedObjectContext {
+        let context = storeContainer.newBackgroundContext()
+        return context
     }
-  }
-
-  public func saveDerivedContext(_ context: NSManagedObjectContext) {
-    context.perform {
-      do {
-        try context.save()
-      } catch let error as NSError {
-        fatalError("Unresolved error \(error), \(error.userInfo)")
-      }
-
-      self.saveContext(self.mainContext)
+    
+    func saveContext(context: NSManagedObjectContext = CoreDataStack.shared.mainContext) throws {
+        var error: Error?
```

Comment:
Optional: Add a guard statement to verify that there are in fact changes on the context to save, otherwise bypass the saving all together.
```suggestion
        guard context.hasChanges else { return }
        var error: Error?
```

Reply:
Implemented suggested change in PR #33